velox/functions/prestosql/window/FirstLastValue.cpp

```diff
@@ -82,9 +83,63 @@ class FirstLastValueFunction : public exec::WindowFunction {
     invalidRows_.applyToSelected([&](auto i) { rowNumbers_[i] = kNullRow; });
   }
 
+  void setRowNumbersRespectNulls(
+      const SelectivityVector& validRows,
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds) {
+    if constexpr (TValue == ValueType::kFirst) {
+      auto rawFrameStarts = frameStarts->as<vector_size_t>();
+      validRows.applyToSelected(
+          [&](auto i) { rowNumbers_[i] = rawFrameStarts[i]; });
+    } else {
+      auto rawFrameEnds = frameEnds->as<vector_size_t>();
+      validRows.applyToSelected(
+          [&](auto i) { rowNumbers_[i] = rawFrameEnds[i]; });
+    }
+  }
+
+  void setRowNumbersIgnoreNulls(
+      const SelectivityVector& validRows,
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds) {
+    auto extractNullsResult = partition_->extractNulls(
+        valueIndex_, validRows, frameStarts, frameEnds, &nulls_);
+
+    auto firstFrame = extractNullsResult.first;
+    auto frameSize = extractNullsResult.second;
+    if (bits::findFirstBit(nulls_->as<uint64_t>(), 0, frameSize) < 0) {
```

Comment:
Would it help to help partition_->extractNulls to return an std::optional which is set only if there are null and unset if there are no nulls?